ENH: GH17054: read_html() handles rowspan/colspan and infers headers

jowens opened the pull request:
- [x] closes #17054
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master -u -- "*.py" | flake8 --diff``
 - [x] whatsnew entry

jreback committed:
DOC: whatsnew typos

jowens committed:
ENH: GH17054: read_html() handles rowspan/colspan and infers headers

jowens committed:
ENH: GH17054: read_html() handles rowspan/colspan and infers headers

gfyoung committed:
TST: Check more error messages in tests (#17075)

toobaz committed:
BUG: Respect dtype when calling pivot_table with margins=True

closes #17013

This fix actually exposed an occurrence of #17035 in an existing test
(as well as in one I added).

Author: Pietro Battiston <me@pietrobattiston.it>

Closes #17062 from toobaz/pivot_margin_int and squashes the following commits:

2737600 [Pietro Battiston] Removed now obsolete workaround
956c4f9 [Pietro Battiston] BUG: respect dtype when calling pivot_table with margins=True

gfyoung committed:
MAINT: Add missing space in parsers.pyx

"2< heuristic" --> "2 < heuristic"

gfyoung committed:
MAINT: Add missing paren around print statement

Stray verbose print statement in parsers.pyx was bare without any parentheses.

jreback committed:
DOC: fix typos in missing.rst

xref #16972

jowens committed:
in python 3, lambdas no longer take tuples as args. thanks pep 3113.

jowens committed:
in python 3, lambdas no longer take tuples as args. thanks pep 3113.

jowens commented:
@gfyoung can you help me reconcile failing tests in the CI checks vs. clean tests on my machine?

```
$ py.test-2.7 pandas/tests/io/test_html.py
============================= test session starts ==============================
platform darwin -- Python 2.7.13, pytest-3.1.3, py-1.4.34, pluggy-0.4.0
rootdir: /Users/jowens/Documents/src/pandas-jowens, inifile: setup.cfg
collected 76 items

pandas/tests/io/test_html.py ............................................................................

========================== 76 passed in 49.13 seconds ==========================
```

gfyoung commented:
@jowens : Judging from your Travis build output, here's what I see:

* You have a style-check error on the Python 2.7 machine.  However, all tests are passing.  That indicates your patch is working for Python 2.7.  You should patch this first to get that build working at least.

* Your changes are failing for Python 3.x, meaning that they are perhaps not compatible with Python 3.  I would suggest that you create a separate environment using Python 3 as your interpreter (I see you are using Python 2), install your branch, and rerun tests to see if you can replicate the failures.

If you can, then continue developing in the Python 3.x environment to get tests passing.  If you can't replicate, then that's going to be tricky to debug, but we can address that *if* we get there.

jowens commented:
Can you point me to the "style-check error"?

gfyoung commented:
The error can be found <a href="https://travis-ci.org/pandas-dev/pandas/jobs/258193317#L1818">here</a>.  BTW, you can also see the error yourself by running the style-check command in the issue PR description:

~~~
git diff upstream/master -u -- "*.py" | flake8 --diff
~~~

jowens commented:
Yeah, I did that before checking in:

```
$ git diff upstream/master -u -- "*.py" | flake8-2.7 --diff
$
```

Not sure what went wrong. I'm on it!

jowens committed:
fixing lint error

jowens committed:
fixing lint error

jowens committed:
in python3, zip does not return a list, so list(zip(...))

jowens committed:
in python3, zip does not return a list, so list(zip(...))

jowens commented:
Cool! @gfyoung what's next?

jorisvandenbossche committed:
DOC: further clean-up null/na changes (#17113)

mroeschke committed:
BUG: Allow pd.unique to accept tuple of strings (#17108)

mroeschke committed:
BUG: Allow Series with same name with crosstab (#16028)

Closes gh-13279

jreback committed:
COMPAT: make sure use_inf_as_null is deprecated (#17126)

closes #17115

jreback committed:
CI: bump version of xlsxwriter to 0.5.2 (#17142)

gfyoung committed:
DOC: Clean up instructions in ISSUE_TEMPLATE (#17146)

FKint committed:
Add missing space to the NotImplementedError's message for compound dtypes (#17140)

jebob committed:
DOC: (de)type the return value of concat (#17079) (#17119)

gfyoung committed:
BUG: Thoroughly dedup column names in read_csv (#17095)

alanyee committed:
DOC: Additions/updates to documentation (#17150)

jreback committed:
ENH: add to/from_parquet with pyarrow & fastparquet (#15838)

jreback committed:
DOC: doc typos, xref #15838

jreback committed:
TST: test for categorical index monotonicity (#17152)

* correctly determine bottleneck version

* tests for categorical index monotonicity

* fix Index.is_monotonic to point to Index.is_monotonic_increasing directly

jbrockmendel committed:
MAINT: Remove non-standard and inconsistently-used imports (#17085)

Commit:
DOC: typos in whatsnew

jreback committed:
DOC: whatsnew 0.21.0 fixes

Commit:
BUG: Fix CSV parsing of singleton list header (#17090)

Closes gh-7757.

jschendel committed:
ENH: Support strings containing '%' in add_prefix/add_suffix (#17151) (#17162)

jorisvandenbossche committed:
REF: repr - allow block to override values that get formatted (#17143)

gfyoung committed:
MAINT: Drop unnecessary newlines in issue template

jbrockmendel committed:
remove direct import of nan

Author: Brock Mendel <jbrockmendel@gmail.com>

Closes #17185 from jbrockmendel/dont_import_nan and squashes the following commits:

ee260b86a [Brock Mendel] remove direct import of nan

jhelie committed:
use == to test String equality (#17171)

deniederhut committed:
ENH: Add warning when setting into nonexistent attribute (#16951)

 closes #7175
 closes #5904

natethedrummer committed:
DOC: added string processing comparison with SAS  (#16497)

jbrockmendel committed:
CLN: remove unused get methods in internals (#17169)

* Remove unused get methods that would raise AttributeError if called

* Remove unnecessary import

mroeschke committed:
TST: Partial Boolean DataFrame Indexing (#17186)

Closes gh-17170

gfyoung committed:
CLN: Reformat docstring for IPython fixture

jbrockmendel committed:
Define Series.plot and Series.hist in class definition (#17199)

toobaz committed:
BUG: support pandas objects in iloc with old numpy versions (#17194)

closes #17193

jbrockmendel committed:
Implement _make_accessor classmethod for PandasDelegate (#17166)

jbrockmendel committed:
Create ABCDateOffset (#17165)

discort committed:
BUG: resample and apply modify the index type for empty Series (#17149)

topper-123 committed:
DOC: Updated NDFrame.astype docs (#17203)

dhimmel committed:
MAINT: Minor touch-ups to GitHub PULL_REQUEST_TEMPLATE (#17207)

Remove leading space from task-list so that tasks aren't nested.

jschendel committed:
CLN: replace %s syntax with .format in core.computation (#17209)

chrisjbillington committed:
Bugfix for multilevel columns with empty strings in Python 2 (#17099)

jorisvandenbossche committed:
CLN/ASV clean-up frame stat ops benchmarks (#17205)

FXocena committed:
BUG: Rolling apply on DataFrame with Datetime index returns NaN (#17156)

dhimmel committed:
CLN: Remove import exception handling (#17218)

Imports should succeed on all versions of Python that pandas supports.

gfyoung committed:
MAINT: Remove extra the's in deprecation messages (#17222)

gfyoung committed:
DOC: Patch docs in _decorators.py

jschendel committed:
CLN: replace %s syntax with .format in pandas.util (#17224)

topper-123 committed:
Add 'See also' sections (#17223)

jbrockmendel committed:
move pivot_table doc-string to DataFrame (#17174)

jbrockmendel committed:
Remove import of pandas as pd in core.window (#17233)

kernc committed:
TST: Move more frame tests to SharedWithSparse (#17227)

toobaz committed:
REF: _get_objs_combined_axis (#17217)

cpcloud committed:
ENH/PERF: Remove frequency inference from .dt accessor (#17210)

* ENH/PERF: Remove frequency inference from .dt accessor

* BENCH: Add DatetimeAccessor benchmark

* DOC: Whatsnew

jbrockmendel committed:
Fix apparent typo in tests (#17247)

mattip committed:
COMPAT: avoid calling getsizeof() on PyPy

closes #17228

Author: mattip <matti.picus@gmail.com>

Closes #17229 from mattip/getsizeof-unavailable and squashes the following commits:

d2623e423 [mattip] COMPAT: avoid calling getsizeof() on PyPy

jschendel committed:
CLN: replace %s syntax with .format in pandas.core.reshape (#17252)

Replaced %s syntax with .format in pandas.core.reshape.  Additionally, made some of the existing positional .format code more explicit.

dhimmel committed:
ENH: Infer compression from non-string paths (#17206)

jschendel committed:
Fix bugs in IntervalIndex.is_non_overlapping_monotonic (#17238)

DGrady committed:
BUG: Fix behavior of argmax and argmin with inf (#16449) (#16449)

Closes #13595

jbrockmendel committed:
CLN: Remove have_pytz (#17266)

Closes gh-17251

jschendel committed:
CLN: replace %s syntax with .format in core.dtypes and core.sparse (#17270)

jbrockmendel committed:
Replace imports of * with explicit imports (#17269)

xref #17234

swyoon committed:
TST: pytest deprecation warnings GH17197 (#17253)

Test parameters with marks are updated according to the updated API of
Pytest.
https://docs.pytest.org/en/latest/changelog.html#pytest-3-2-0-2017-07-30
https://docs.pytest.org/en/latest/parametrize.html

Winand committed:
Handle more date/datetime/time formats (#15871)

zzgao committed:
DOC: add example on json_normalize (#16438)

TomAugspurger committed:
BUG: Have object dtype for empty Categorical.categories (#17249)

* BUG: Have object dtype for empty Categorical ctor

Previously we had a `Float64Index`, which is inconsistent with, e.g., the
regular Index constructor.

* TST: Update tests in multi for new return

Previously these relied worked around the return type by wrapping list-likes
in `np.array` and relying on that to cast to float. These workarounds are no
longer nescessary.

* TST: Update union_categorical tests

This relied on `NaN` being a float and empty being a float. Not a necessary
test anymore.

* TST: set object dtype

jschendel committed:
CLN: replace %s syntax with .format in pandas.tseries (#17290)

jreback committed:
TST: parameterize consistency tests for rolling/expanding windows (#17292)

tacaswell committed:
FIX: define `DataFrame.items` for all versions of python (#17214)

TomAugspurger committed:
PERF: Update ASV publish config (#17293)

Stricter cutoffs for considering regressions

[ci skip]

yosukeBaya4 committed:
DOC: Expand docstrings for head / tail methods (#16941)

gfyoung committed:
MAINT: Use set literal for unsupported + depr args

Initializes unsupported and deprecated argument sets with set literals instead of the set constructor in pandas/io/parsers.py, as the former is slightly faster than the latter.

gfyoung committed:
DOC: Add proper docstring to maybe_convert_indices

Patches several spelling errors and expands current doc to a proper doc-string.

matagus committed:
DOC: Improving docstring of take method (#16948)

TomAugspurger committed:
BUG: Fixed regex in asv.conf.json (#17300)

In https://github.com/pandas-dev/pandas/pull/17293 I messed up the syntax. I
used a glob instead of a regex. According to the docs at
http://asv.readthedocs.io/en/latest/asv.conf.json.html#regressions-thresholds we
want to use a regex. I've actually manually tested this change and verified that
it works.

[ci skip]

jbrockmendel committed:
Remove unnecessary usage of _TSObject (#17297)

Commit:
BUG: clip should handle null values

closes #17276

Author: Michael Gasvoda <mgasvoda@mercatus.gmu.edu>
Author: mgasvoda <mgasvoda01@gmail.com>

Closes #17288 from mgasvoda/master and squashes the following commits:

a1dbdf293 [mgasvoda] Merge branch 'master' into master
9333952c2 [Michael Gasvoda] Checking output of tests
4e0464eaf [Michael Gasvoda] fixing whatsnew text
c44204080 [Michael Gasvoda] formatting fixes
7e2367879 [Michael Gasvoda] formatting updates
781ea724a [Michael Gasvoda] whatsnew entry
d9627fe4c [Michael Gasvoda] adding clip tests
9aa0159e9 [Michael Gasvoda] Treating na values as none for clips

Commit:
BUG: fillna returns frame when inplace=True if value is a dict (#16156) (#17279)

toobaz committed:
CLN: Index.append() refactoring (#16236)

jreback committed:
DEPS: set min versions (#17002)

closes #15206, numpy >= 1.9
closes #15543, matplotlib >= 1.4.3
scipy >= 0.14.0

jschendel committed:
CLN: replace %s syntax with .format in core.tools, algorithms.py, base.py (#17305)

Commit:
BUG: Fix strange behaviour of Series.iloc on MultiIndex Series (#17148) (#17291)

gfyoung committed:
DOC: Add module doc-string to tseries/api.py

gfyoung committed:
MAINT: Clean up docs in pandas/errors/__init__.py

jschendel committed:
CLN: replace %s syntax with .format in missing.py, nanops.py, ops.py (#17322)

Replaced %s syntax with .format in missing.py, nanops.py, ops.py. Additionally, made some of the existing positional .format code more explicit.

jbrockmendel committed:
Make pd.Period immutable (#17239)

Commit:
Bug: groupby multiindex levels equals rows (#16859)

closes #16843

ivybae committed:
BUG: Cannot use tz-aware origin in to_datetime (#16842)

closes #16842

Author: step4me <prosikeffect@gmail.com>

Closes #17244 from step4me/step4me-feature and squashes the following commits:

09d051d48 [step4me] BUG: Cannot use tz-aware origin in to_datetime (#16842)

jbrockmendel committed:
Replace usage of total_seconds compat func with timedelta method (#17289)

cbertinato committed:
CLN: replace %s syntax with .format in core/indexing.py (#17357)

Progress toward issue #16130. Converted old string formatting to new string formatting in core/indexing.py.

gfyoung committed:
DOC: Point to dev-docs in issue template (#17353)

[ci skip]

chris-b1 committed:
CLN: remove total_seconds compat from json (#17341)

jschendel committed:
CLN: Move test_intersect_str_dates (#17366)

Moves test_intersect_str_dates from tests/indexes/test_range.py to tests/indexes/test_base.py.

gfyoung committed:
BUG: Respect dups in reindexing CategoricalIndex (#17355)

When the indexer is identical to the elements.
We should still return duplicates when the indexer
contains duplicates.

Closes gh-17323.

jbrockmendel committed:
Unify Index._dir_* with Series implementation (#17117)

jowens commented:
@gfyoung any chance of moving this forward? I'd prefer not to have a private build of Pandas for the rest of my life.

gfyoung commented:
@jowens : Sorry about that!  My inbox is very full most of the time, so I think I just missed your comment about a month ago.

Let's do this:
* Rebase onto `master` (and fix the merge conflict) to see if tests still pass
* Ping @TomAugspurger for further review of the PR

jowens commented:
@gfyoung FWIW the merge conflict is only "whatsnew"

jowens commented:
@gfyoung because I *really* don't want to get this wrong and have a commit that's gonna be enormous, can you sketch out how I "rebase onto master" properly?

gfyoung commented:
@jowens: True, but it's been a month since you've run the tests on Travis, Appveyor, etc.  I just want to make sure that tests still pass.

gfyoung commented:
> @gfyoung because I really don't want to get this wrong and have a commit that's gonna be enormous, can you sketch out how I "rebase onto master" properly?

I would recommend to try using GitHub's UI editor for resolving the merge conflict.  If that doesn't work out, I can give you further suggestions, but they will be all local (and probably) less visual.

gfyoung commented:
Don't worry about rebasing onto master yet.  Let's make sure this PR is merge-able first.

jowens committed:
Merge branch 'master' into read_html_with_colspan_rowspan

jowens commented:
@gfyoung OK, looks like it's running and conflict-free. Stay tuned!

toobaz committed:
BUG: make order of index from pd.concat deterministic (#17364)

closes #17344

jbrockmendel committed:
Fix typo that causes several NaT methods to have incorrect docstrings (#17327)

jowens commented:
@TomAugspurger can you help me take the next step with this PR? Thanks @gfyoung!

cbertinato committed:
CLN: replace %s syntax with .format in io/formats/format.py (#17358)

Progress toward issue #16130. Converted old string formatting to new string formatting in io/formats/format.py.

TomAugspurger committed:
PKG: Added pyproject.toml for PEP 518 (#16745)

Declaring build-time requirements: https://www.python.org/dev/peps/pep-0518/

TomAugspurger commented:
I won't have a chance to review this in detail for a while, so if you think
it's good then feel free to merge @gfyoung.

On Tue, Aug 29, 2017 at 3:14 PM, John Owens <notifications@github.com>
wrote:

> @TomAugspurger <https://github.com/tomaugspurger> can you help me take
> the next step with this PR? Thanks @gfyoung <https://github.com/gfyoung>!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/17089#issuecomment-325787329>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIumSd6yu0bcha3VfFOCHko-9Fx8Aks5sdHErgaJpZM4Okn1p>
> .
>

iuliakhomenko committed:
DOC: Update Overview page in documentation (#17368)

* Update Overview page in documentation

* DOC Revise Overview page

* DOC Make further revisions in Overview webpage

* Update overview.rst

Remove references to Panel

TomAugspurger committed:
API: Have MultiIndex consturctors always return a MI (#17236)

* API: Have MultiIndex constructors return MI

This removes the special case for MultiIndex constructors returning
an Index if all the levels are length-1. Now this will return a
MultiIndex with a single level.

This is a backwards incompatabile change, with no clear method for
deprecation, so we're making a clean break.

Closes #17178

* fixup! API: Have MultiIndex constructors return MI

* Update for comments

cbertinato committed:
CLN: replace %s syntax with .format in io/formats/css.py, excel.py, printing.py, style.py, and terminal.py (#17387)

Progress toward issue #16130. Converted old string formatting to new string formatting in io/formats/css.py, excel.py, printing.py, style.py, and terminal.py

sylviawhoa committed:
BUG: not correctly using OrderedDict in test_series_apply (#17384)

in Python versions <3.6 this syntax will result in an unordered dic

jbrockmendel committed:
Remove boxplot from _dataframe_apply_whitelist (#17381)

mroeschke committed:
API: Localize Series when calling to_datetime with utc=True (#6415) (#17109)

jschendel committed:
TST: Enable tests in test_tools.py (#17405)

Enabled tests that currently aren't running.  Small fix to make sure all tests pass.  Verified that the raised messages match expectations for TestToDatetimeUnit::test_frame.

topper-123 committed:
TST: remove tests and docs for legacy (pre 0.12) hdf5 support (#17404)

jbrockmendel committed:
Tslib unused (#17402)

topper-123 committed:
DOC: Cleaned references to pandas <v0.12 in docs (#17375)

jbrockmendel committed:
Remove unused _day and _month attrs (#17431)

closes #17429

topper-123 committed:
DOC: Clean-up references to v12 to v14 (both included) (#17420)

s-weigand committed:
BUG: Plotting Timedelta on y-axis #16953 (#17430)

* implemented fix for GH issue #16953

* added tests for fix of issue #16953

* changed comments for git issue to pandas style GH#

* changed linelength in tests, so all lines are less than 80 characters

* added whatsnew entry

* swaped conversion and filtering of values, for plot to also work with object dtypes

* refomated code, so len(line) < 80

* changed whatsnew with timedelta and datetime dtypes

* added support for datetimetz and extended tests

* added reason to pytest.mark.xfail

jreback committed:
COMPAT: handle pyarrow deprecation of timestamps_to_ms in .from_pandas with pyarrow < 0.6.0 (#17447)

closes #17438

topper-123 committed:
DOC/TST: Add examples to MultiIndex.get_level_values + related changes (#17414)

jowens commented:
@TomAugspurger @gfyoung could either one of you guys please take a look?

gfyoung commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -124,14 +124,13 @@ Other Enhancements
 - :func:`DataFrame.items` and :func:`Series.items` is now present in both Python 2 and 3 and is lazy in all cases (:issue:`13918`, :issue:`17213`)
 
-
-
```
Add these lines back.

gfyoung replied:
Also, your `whatsnew` has a merge conflict, so you definitely need to fix that before we merge this.

gfyoung commented on pandas/io/html.py:
```diff
@@ -197,29 +199,32 @@ def parse_tables(self):
         ----------
-        rows : iterable of node-like
-            A list of row elements.
+        doc : tree-like
```
As tree-like is not a Python data-type, we should re-specify this.

gfyoung replied:
@TomAugspurger : If you think "node-like" or "tree-like" is OK for this documentation, then we can ignore this comment.  However, IMO we should have proper data types in the doc-strings.

gfyoung commented on pandas/io/html.py:
```diff
@@ -236,48 +241,58 @@ def _text_getter(self, obj):
 
         Parameters
         ----------
         obj : node-like
```
Let's take advantage of this time to re-specify this as a proper data type if possible.

gfyoung commented on pandas/io/html.py:
```diff
@@ -236,48 +241,58 @@ def _text_getter(self, obj):
         obj : node-like
+            A DOM node.
+
+        tag : string
```
string --> str

gfyoung commented on pandas/io/html.py:
```diff
@@ -236,48 +241,58 @@ def _text_getter(self, obj):
-        columns : list of node-like
-            These are the elements of each row, i.e., the columns.
+        boolean
+            Does the object match tag 'tag'?
```
Just say "boolean indicating the object matches the tag `tag`"

gfyoung commented on pandas/io/html.py:
```diff
@@ -236,48 +241,58 @@ def _text_getter(self, obj):
 
-        match : str or regular expression
-            The text to search for in the DOM tree.
+        tag : string
```
string --> str

gfyoung commented on pandas/io/html.py:
```diff
@@ -236,48 +241,58 @@ def _text_getter(self, obj):
         ----------
-        doc : tree-like
-            The DOM from which to parse the table element.
+        obj : node-like
```
Let's try to re-specify as something other than node-like.

jowens replied:
I totally copied "node-like" from other parts of the documentation. Suggestion of what I should use?

gfyoung replied:
@jowens : See discussion <a href="https://github.com/pandas-dev/pandas/pull/17089#discussion_r137319069">above</a>.  I don't want to buck convention if this has been the thing to use.  However, the best thing you can do is try to determine the data-type yourself when you call the function.  For example, add a `print(type(obj))` in your implementation to see what it is.

gfyoung replied:
Just provide the data-type that you get when you add `print(type(obj))` to the implementation of the function.  I'm not entirely sure what it is myself, hence why I suggested you do that to get an exact answer.

gfyoung commented on pandas/io/html.py:
```diff
@@ -236,48 +241,58 @@ def _text_getter(self, obj):
+        Returns
+        -------
+        boolean
+            Does the object contain tag 'tag'?
```
Use the same boolean description as above: "Boolean indicating whether..."

gfyoung commented on pandas/io/html.py:
```diff
@@ -236,48 +241,58 @@ def _text_getter(self, obj):
+
+        Parameters
+        ----------
+        obj : node-like
```
Let's take advantage of this time to re-specify this as a proper data type if possible.

gfyoung commented on pandas/io/html.py:
```diff
@@ -346,40 +361,149 @@ def _build_doc(self):
+
+        Parameters
+        ----------
+        table_html : node-like
```
Let's take advantage of this time to re-specify this as a proper data type if possible.

gfyoung commented on pandas/io/html.py:
```diff
@@ -346,40 +361,149 @@ def _build_doc(self):
+
+        Returns
+        -------
+        header, body, footer
```
Say "tuple of (header, body, footer)"

jowens committed:
Merge branch 'master' into read_html_with_colspan_rowspan

jowens committed:
documentation changes only

jowens committed:
documentation changes only

jowens committed:
Merge branch 'read_html_with_colspan_rowspan' of github.com:jowens/pandas into read_html_with_colspan_rowspan

jowens commented:
OK @gfyoung next step?

gfyoung commented on pandas/io/html.py:
```diff
@@ -197,36 +199,39 @@ def parse_tables(self):
         ----------
-        rows : iterable of node-like
-            A list of row elements.
+        doc : parser object
```
I think you still need to supply the actual data type of the input i.e. what is the output of type(`doc`) when used in this function?  Just add a line in the function that prints out the `type` and recompile `pandas` to include it.  Then see what it is and then add that to the documentation.

jowens replied:
OK. It depends on what parser you use, though (BeautifulSoup or lxml). It's the BS4 or lxml parser object. I will check exact types.

gfyoung replied:
Arghhh!  I wish you had told me this sooner. :cry:

gfyoung replied:
For future reference, we are *generally* very picky about the data types in the doc-string.  However, as the data-type is dependent on an external factor (i.e. engine), we drop the data type and immediately dive into the description of the variable.

gfyoung replied:
Here is what I mean by that for example:

BEFORE
~~~python
row : node-like
     A row in the HTML table
~~~

AFTER
~~~python
row : a row in the HTML table
~~~

jowens replied:
Sorry. Didn't know that until I dived into the documentizing. So right now I am consistent about this kind of docstring:

```
        rows : list of HTML row elements
            A list of row elements of a table, usually <tr> or <th> elements.
```

but the `row element` here is also dependent on the parser (it's either a BS4 or lxml row element). This does seem like the right way to describe it though. (I left the documentation structured as is in the way it described subclassed functions; there is an virtual function, and that's what's documented, and then subclassed functions specific to the parser that are not documented.)

So: 

- Would you like me to back out the changes I made (the ones like the above) and redo them as you suggest?
- Based on what you suggest, `parser object` does seem like the right way to describe it ...?

gfyoung replied:
@jowens : Instead of providing a "data-type" like `parser object`, go right into describing the data input, so in the example you provided above, do this:

~~~
rows : a list of row elements of a table, usually <tr> or <th> elements.
~~~

jowens replied:
OK. For where I have "parser object", here's the actual type: `<class 'bs4.BeautifulSoup'>`. Suggestion for what to put instead of "parser object"?

gfyoung replied:
Just do this:

~~~python
doc : the DOM from which to parse the table element.
~~~

jowens replied:
Many thanks. Next round pushed. You are a very patient person.

jowens committed:
documentation changes only

jowens committed:
documentation changes only

pep8speaks commented:
Hello @jowens! Thanks for updating the PR.

 - In the file [`pandas/io/html.py`](https://github.com/pandas-dev/pandas/blob/5a38278f1858452fa9a3b8953199b7e7c1018441/pandas/io/html.py), following are the PEP8 issues :

> [Line 299:80](https://github.com/pandas-dev/pandas/blob/5a38278f1858452fa9a3b8953199b7e7c1018441/pandas/io/html.py#L299): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 379:80](https://github.com/pandas-dev/pandas/blob/5a38278f1858452fa9a3b8953199b7e7c1018441/pandas/io/html.py#L379): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 381:80](https://github.com/pandas-dev/pandas/blob/5a38278f1858452fa9a3b8953199b7e7c1018441/pandas/io/html.py#L381): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)

jowens committed:
documentation changes only, limited to 80 cols

jowens committed:
documentation changes only, limited to 80 cols

jowens commented:
Hit me with more changes, @gfyoung!

gfyoung commented on pandas/io/html.py:
```diff
@@ -197,37 +199,38 @@ def parse_tables(self):
         -------
-        data : list of list of strings
+        tables : list of HTML table elements
+            A list of <table> elements to be parsed into raw data.
```
This is repetitive.  Just put your description in the line above.

gfyoung commented on pandas/io/html.py:
```diff
@@ -236,104 +239,104 @@ def _text_getter(self, obj):
-        columns : list of node-like
-            These are the elements of each row, i.e., the columns.
+        boolean
+            boolean indicating if the object is equal to tag 'tag'
```
Write:

~~~python
is_tag_equal : boolean
    boolean indicating if the object is equal to tag 'tag'
~~~

gfyoung commented on pandas/io/html.py:
```diff
@@ -342,44 +345,149 @@ def _build_doc(self):
+        elements
+        body : list of rows, each of which is a list of parsed body elements
+        footer : list of rows, each of which is a list of parsed footer
+        elements
```
Visually, this formatting looks weird.  Can you indent any subsequent lines in your description and avoid lone words on a line.  For example:

~~~python
header : list of rows, each of which is
         a list of header elements
~~~

gfyoung commented on pandas/io/html.py:
```diff
@@ -342,44 +345,149 @@ def _build_doc(self):
+                footer_rows = self._extract_tr(
+                    self._extract_tfoot(table_html)[0])
+        else:
+            # otherwise we need to split the body into header/body/foot
```
otherwise --> Otherwise
Add period at end.

gfyoung commented on pandas/io/html.py:
```diff
@@ -342,44 +345,149 @@ def _build_doc(self):
+        body = self._expand_colspan_rowspan(body_rows, fill_rowspan=True)
+        footer = self._expand_colspan_rowspan(footer_rows, fill_rowspan=False)
+        # the below line is lifted from _parse_raw_tfoot. Not sure what it
+        # does.
```
Add newline above your comment.  Rewrite so that "does" is not alone in this line.

jowens committed:
more documentation edits

jowens committed:
more documentation edits

jowens committed:
minor documentation edits

jowens committed:
minor documentation edits

gfyoung commented on pandas/io/html.py:
```diff
@@ -236,104 +238,104 @@ def _text_getter(self, obj):
-            mutliple tables on a page.
+        Returns
+        -------
+        boolean
```
Similar to `is_tag_equal` above, provide a variable name here and then the data type.

gfyoung commented on pandas/tests/io/test_html.py:
```diff
@@ -640,6 +639,118 @@ def test_different_number_of_rows(self):
         tm.assert_frame_equal(expected, res)
 
+    def test_colspan_rowspan_are_1(self):
+        expected = """<table>
```
Reference issue number above

gfyoung replied:
Do this for all of the tests that you added.

jowens committed:
better return type explanation in code, added issue number to tests

jowens committed:
better return type explanation in code, added issue number to tests

jowens commented:
@gfyoung @TomAugspurger Next steps?

gfyoung commented on pandas/io/html.py:
```diff
@@ -197,37 +199,37 @@ def parse_tables(self):
+        Raises
+        ------
+        ValueError
+            * If `match` does not match any text in the document.
```
Let's move this onto the previous line with a colon after `ValueError`

gfyoung commented on pandas/io/html.py:
```diff
@@ -342,44 +344,150 @@ def _build_doc(self):
-        header = self._parse_raw_thead(table)
-        body = self._parse_raw_tbody(table)
-        footer = self._parse_raw_tfoot(table)
+    def _build_table(self, table_html):
```
Add a brief doc-string here.

gfyoung commented:
@jowens : Sorry for not looking at this sooner.  I made some additional comments, but it's definitely well on its way to getting merged.  Feel free to ping us again as soon as you finish addressing my comments.

jowens committed:
cleaning up legacy documentation issues

jowens committed:
cleaning up legacy documentation issues

gfyoung commented on pandas/io/html.py:
```diff
@@ -211,12 +217,11 @@ def _parse_tables(self, doc, match, attrs):
         ------
-        ValueError
-            * If `match` does not match any text in the document.
+        ValueError : If `match` does not match any text in the document.
```
Remove the "if"

jowens commented:
@gfyoung I expect no issues with the build since it's documentation-changes-only. FWIW these are documentation issues that predate any of my changes (but I don't mind cleaning them up).

jowens committed:
remove 'if'

jowens committed:
remove 'if'

gfyoung commented:
@jowens : Understood.  I suggested them because those docs are showing up in your diff, so they're fair game as far as I'm concerned. :smile:

gfyoung commented on pandas/io/html.py:
```diff
@@ -342,44 +349,150 @@ def _build_doc(self):
+        footer : list of rows, each of which is a list of parsed
+                 footer elements
+        """
+        header_rows = []
```
Nit: Add newline above this one.

gfyoung commented on pandas/io/html.py:
```diff
@@ -342,44 +349,150 @@ def _build_doc(self):
+        header_rows = []
+        body_rows = []
+        footer_rows = []
+        # first, are there thead and tbody elements in the table?
```
Nit: add newline above this one.

gfyoung commented on pandas/io/html.py:
```diff
@@ -342,44 +349,150 @@ def _build_doc(self):
+                self._contains_tag(table_html, 'tbody')):
+            header_rows = self._extract_tr(self._extract_thead(table_html)[0])
+            body_rows = self._extract_tr(self._extract_tbody(table_html)[0])
+            if self._contains_tag(table_html, 'tfoot'):
```
Nit: add newline above this one.

gfyoung commented on pandas/io/html.py:
```diff
@@ -342,44 +349,150 @@ def _build_doc(self):
+                      self._extract_td(body_rows[-1])):
+                # this row should be a footer row, move it from body to footer
+                footer_rows.insert(0, body_rows.pop())
+        header = self._expand_colspan_rowspan(header_rows, fill_rowspan=False)
```
Nit: add newline above this one.

gfyoung commented on pandas/io/html.py:
```diff
@@ -342,44 +349,150 @@ def _build_doc(self):
+        res : list of rows, each of which is a list of elements in that row,
+            respecting colspan/rowspan
+        """
         res = []
```
Nit: add newline above this one.

jowens committed:
newlines for clarity

jowens committed:
newlines for clarity

gfyoung reviewed (approved):
LGTM.  Thanks for the changes!

jowens commented:
@gfyoung Any idea why Travis CI failed here? It passed before and all I changed recently was documentation.

gfyoung commented:
@jowens : This was handled in in #17588.  Rebase on `master` and push again.

jowens committed:
Merge branch 'read_html_with_colspan_rowspan' of github.com:jowens/pandas into read_html_with_colspan_rowspan

jowens commented:
@TomAugspurger Next steps?

jowens commented:
@TomAugspurger @gfyoung What can I do to move this forward?

gfyoung commented:
@TomAugspurger : Can you take a look at this?  I've already approved, and @jowens and I are awaiting your comments before proceeding.

TomAugspurger commented on pandas/io/html.py:
```diff
@@ -17,10 +17,10 @@
                            raise_with_traceback, binary_type)
 from pandas import Series
-from pandas.core.common import AbstractMethodError
+from pandas.core.common import (AbstractMethodError, flatten)
```
Don't need the parens

TomAugspurger commented on pandas/io/html.py:
```diff
@@ -194,40 +196,45 @@ def __init__(self, io, match, attrs, encoding):
-            defined by subclasses.
+        match : str or regular expression
+            The text to search for in the DOM tree.
+
```
You can remove these blank lines between parameters

TomAugspurger commented on pandas/io/html.py:
```diff
@@ -342,44 +349,155 @@ def _build_doc(self):
+
+        Returns
+        -------
+        tuple of (header, body, footer)
```
you can remove this line (multiple returns are always tuples)

TomAugspurger commented on pandas/io/html.py:
```diff
@@ -342,44 +349,155 @@ def _build_doc(self):
-        footer = self._parse_raw_tfoot(table)
+    def _build_table(self, table_html):
+        header, body, footer = self._parse_raw_thead_tbody_tfoot(table_html)
+        # the above "footer" actually produces a footer. The below "footer"
```
What's the plan here? Do we have a GH issue open for replacing the old with the new? i.e., what does the person reading the code 3 years from now need to know whether this has been resolved?

TomAugspurger commented on pandas/io/html.py:
```diff
@@ -342,44 +349,155 @@ def _build_doc(self):
+            while all(self._equals_tag(t, 'th') for t in
+                      self._extract_td(body_rows[-1])):
+                # this row should be a footer row, move it from body to footer
+                footer_rows.insert(0, body_rows.pop())
```
This seems to be untested.

TomAugspurger commented on pandas/io/html.py:
```diff
@@ -342,44 +349,155 @@ def _build_doc(self):
+            # splitting criterion: if all tags within a row are th, it's part
+            # of the header/footer
+            while all(self._equals_tag(t, 'th') for t in
+                      self._extract_td(body_rows[0])):
```
What happens with a table of all "headers". Does the condition need to be `while body_rows and all(...)` (is it possible to get to this situation, and do you have a test for it)?

TomAugspurger commented on pandas/io/html.py:
```diff
@@ -342,44 +349,155 @@ def _build_doc(self):
+                            for col in extracted_row]
+            # expand cols using col_colspans
+            # maybe this can be done with a list comprehension, dunno
+            cols = list(zip(
```
I think this is equivalent to

```python
list(chain.from_iterable([text] * nc for text, nc in zip(cols_text, col_colspans)))
```

which I find a bit more readable. Can you confirm if that works? (import `chain` from itertools).

TomAugspurger replied:
Hmm, reading a bit further, maybe not? But I think that's close. Could you post a sample of `col_colspans`, `col_rowspans`, and `cols_texts`? Are they all just flat lists of ints / strs?

TomAugspurger commented on pandas/io/html.py:
```diff
@@ -443,6 +540,33 @@ def _parse_tables(self, doc, match, attrs):
+        return element.find_all('tr')
+
+    def _extract_th(self, element):
+        return element.find_all('th')
```
Untested?

TomAugspurger commented on pandas/io/html.py:
```diff
@@ -585,24 +715,24 @@ def _build_doc(self):
         res = []
         if thead:
-            trs = self._parse_tr(thead[0])
+            trs = self._extract_tr(thead[0])
```
Untested?

TomAugspurger commented on pandas/io/html.py:
```diff
@@ -873,7 +1003,13 @@ def read_html(io, match='.+', flavor=None, header=None, index_col=None,
+    the header, otherwise the function attempts to find the header within
+    the body (by putting rows with only ``<th>`` elements into the header).
+
+        .. versionadded:: 0.21.0
```
This could be a `versionmodified`, though it may not be necessary.

TomAugspurger reviewed (commented):
Overall things look pretty good. I don't completely understand the logic in `_expand_colspan_rowspan`, and there were several lines that seem to be untested (perhaps they're dead code now?).

Could you attempt to refactor those `lmaps` to use list-comprehensions and `chain`? Could you add some more tests for

- table with just a `thead`, but no body
- table with just a `tbody`, but each row is a `th`

And cases where `rowspan` / `colspan` vary for each of those.

jreback commented:
needs to be rebased

jowens commented:
@TomAugspurger made a lot of good suggestions that I simply haven't had time to get to yet. Is there an impending deadline here?

jreback commented:
@jowens nope. just pinging on older pr's. rebase / fixup when you can.

ostrokach commented:
@jowens I just tried out this branch for extracting tables from PubMed Central XML files, and and it works much much better than the current `read_html` implementation (which was basically unusable)! Thank you for your work guys! Hope this can make it into the master branch soon :).

jreback commented:
can you ``git merge master`` resolve conflicts and push again (and move whatsnew to 0.23.0)

jowens commented:
@TomAugspurger's suggestions are still all good ones, but I haven't had a full day where I can sit down and do them.

wkerzendorf commented:
great work! can't wait for it to get merged

jowens commented:
I'll repeat that "@TomAugspurger's suggestions are still all good ones, but I haven't had a full day where I can sit down and do them." :)

TomAugspurger commented:
No rush, we're patient :)

On Thu, Apr 19, 2018 at 1:31 PM, John Owens <notifications@github.com>
wrote:

> I'll repeat that "@TomAugspurger <https://github.com/TomAugspurger>'s
> suggestions are still all good ones, but I haven't had a full day where I
> can sit down and do them." :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/17089#issuecomment-382836750>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIl6-no2ugOPI4Jct6JpVcLw7hqRyks5tqNgVgaJpZM4Okn1p>
> .
>

cancan101 commented:
any updates on this PR?

adamhooper commented:
@jowens Hi! This feature is at the top of my backlog this week. So effectively, _I_ have a full day where I can sit down and incorporate @TomAugspurger's suggestions. Would that be okay with you?

jowens commented:
Yes PLEASE! I would be delighted (and so would the other folks in this thread).

WillAyd commented:
Superseded by #21487
